plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/MonitorInfo.java

```diff
@@ -0,0 +1,174 @@
+        final Map<String, String> systemMetadata = service.provider().metadata();
+
+        final String key = toKey(service);
+        final Bundle oldBundle = infoBundles.get(key);
+        final JsonObject systemData = (oldBundle == null) ? null : oldBundle.systemData;
+
+        final Bundle newBundle = new Bundle(
+            systemName,
+            service.name(),
+            systemMetadata,
+            serviceMetadata,
+            systemData,
+            inventoryId
+        );
+
+        infoBundles.put(key, newBundle);
+    }
+
+    /**
+     * Store system data for the given service. Any system data previously
+     * stored for this system will be overwritten.
+     *
+     * @param service An Arrowhead Framework service.
+     * @param data    System data to be stored.
+     */
+    public void putSystemData(final ServiceRecord service, final JsonObject data) {
+
+        Objects.requireNonNull(service, "Expected service");
+        Objects.requireNonNull(data, "Expected system data");
+
+        final String key = toKey(service);
+        final SystemRecord provider = service.provider();
+        final String systemName = provider.name();
+
+        final Map<String, String> systemMetadata = provider.metadata();
+        final Map<String, String> serviceMetadata = service.metadata();
+        final String serviceDefinition = service.name();
+        final Bundle oldBundle = infoBundles.get(key);
+        final String inventoryId = (oldBundle == null) ? null : oldBundle.inventoryId;
+
+        final Bundle newBundle = new Bundle(
+            systemName,
+            serviceDefinition,
+            systemMetadata,
+            serviceMetadata,
+            data,
+            inventoryId
+        );
+
+        infoBundles.put(key, newBundle);
+    }
+
+    /**
+     * @param systemName Name of an Arrowhead Framework system. May be null if
+     *                   {@code metadata} is present.
+     * @param metadata   Metadata describing an Arrowhead Framework system. May
+     *                   be null if {@code systemName} is present.
+     * @return A list containing all stored monitor data for the system that
+     * corresponds to the given arguments.
+     */
+    public List<Bundle> getSystemInfo(final String systemName, final Map<String, String> metadata) {
+
+        if (systemName == null && metadata == null) {
+            throw new IllegalArgumentException("Either system name or metadata must be present.");
+        }
+
+        final List<Bundle> result = new ArrayList<>();
+        for (final Bundle bundle : infoBundles.values()) {
+            final boolean namesMatch = systemName == null || systemName.equals(bundle.systemName);
+            final boolean metadataMatch = bundle.matchesSystemMetadata(metadata);
+            if (namesMatch && metadataMatch) {
+                result.add(bundle);
+            }
+        }
+        return result;
+    }
+
+    public static class Bundle {
```

Comment:
why public if its only constructor is not public?

Reply:
Instances of the Bundle class were used outside of the class MonitorInfo, but they were only created inside MonitorInfo. However, I've refactored the code now and put the inner class in its own file.